### PR TITLE
detect URLs in rich console and make them clickable

### LINKED ISSUE #421: Auto-URL for rich console
@NathanRichard maybe you want to help me with this.

In https://github.com/chriskiehl/Gooey/commit/f581884fbf86ecb0634c91312cb9dab86c0d2281 and https://github.com/chriskiehl/Gooey/pull/417 we (well @chriskiehl  ) added the auto-URL, so the console identifies a URL as a clickable thing.

This doesn't work for rich controls. So the following code identifies the link if `richtext_controls=False` but not if it's `True`.

I tried to add `| wx.TE_AUTO_URL | wx.TE_RICH` to https://github.com/chriskiehl/Gooey/blob/1.0.3-release/gooey/gui/components/widgets/richtextconsole.py#L41 but it's not enough, things still don't work. 

```
import sys
from gooey import Gooey, GooeyParser

@Gooey(richtext_controls=True,auto_start=True)
def main():
    _ = GooeyParser().parse_args(sys.argv[1:])
    print ('please visit http://google.com')

if __name__ == '__main__':
    main()
```

### LINKED ISSUE #427: Auto-URL is confused by wide characters
related to #404 

```
import sys
from gooey import Gooey, GooeyParser

@Gooey(richtext_controls=False,auto_start=True)
def main():
    _ = GooeyParser().parse_args(sys.argv[1:])
    print ('please 📧📧📧📧 visit http://google.com because life is pretty')

if __name__ == '__main__':
    main()
```

It detects the URL but cliking it tries to go to `http://google.com%20bec` 

So the four extra-wide characters caused the URL to end (and I suspect also: start) 4 character too late. 

I imagine it's something with `event.URLStart:event.URLEnd` counting bytes instead of characters or vice versa,  in  https://github.com/chriskiehl/Gooey/blob/1202f13a5c916a5a4c07590a8c55b523de7cf606/gooey/gui/components/console.py#L43 

tested in python 3 only so far.

### PR DESCRIPTION
**Pull request by eladeyal-intel**
Closes #421 - we detect URLs using a regex and applying the URL style.

Also closes #427 becasue it's the same line

### CODE REVIEW

**Comment by eladeyal-intel** [comment]
Wait... Don't merge

File: gooey/gui/components/widgets/richtextconsole.py [hunk 1]
```diff
@@ -59,6 +61,27 @@ def __init__(self, parent):
             # NB : we use a default parameter to force the evaluation of the binding
             self.actionsMap[escSeq] = lambda bindedColor=wxcolor: self.BeginTextColour(bindedColor)
 
+    def AppendText2(self, content):
```
Comment (chriskiehl) [inline gooey/gui/components/widgets/richtextconsole.py]: Is there a better name that AppendText2?

File: gooey/gui/components/widgets/richtextconsole.py [hunk 2]
```diff
@@ -59,6 +61,27 @@ def __init__(self, parent):
...
+        """Capture URLs and make them blue, underlined, and clickable."""
+        while True:
```
Comment (chriskiehl) [inline gooey/gui/components/widgets/richtextconsole.py]: I don't dig this while loop + mutation of the content variable. 

As a possible alternative, if you expand your regex's capture group to include the whole match, you can simplify this whole thing to a single forloop + if statement

```python 
url_pattern = r'\b((?:file|https?)://[^\s<>]*)'
```

>note that we treat the inner scheme subgroup as non-capturing 

then re.split will give you stable pairs of [plaintext, url]

```python
re.split(r, 'http://www.example.com')
>>> ['', 'http://www.example.com', '']
re.split(r, 'Hello world!\n\nhttp://www.example.com')
>>> ['Hello world!\n\n', 'http://www.example.com', '']
re.split(r, 'file://www.example.com')
>>> ['', 'file://www.example.com', '']
re.split(r, 'some text file://www.example.com some more text http://www.example.com http://www.example.com')
>>> ['some text ', 'file://www.example.com', ' some more text ', 'http://www.example.com', ' ', 'http://www.example.com', '']
```

Combine this with `itertools.grouper` and you can iterate over the pairs directly

```python
for plaintext, url in grouper(re.split(pattern, content), 2):
    self.WriteText(plaintext)
    if url:
        self.BeginTextColour(self.url_colour)
        self.BeginUnderline()
        self.BeginURL(url)
        self.WriteText(url)
        self.EndURL()
        self.EndUnderline()
        self.EndTextColour()
```

Comment (eladeyal-intel) [reply gooey/gui/components/widgets/richtextconsole.py]: `grouper` is not actually in itertools, it's just a "recipe" in https://docs.python.org/3.7/library/itertools.html#itertools-recipes ; and it needs either `izip_longest` or `zip_longest` (for Python 2 vs 3) which seems to me like a hassle. I hope the updated version is more to your liking.

File: gooey/gui/components/widgets/richtextconsole.py [hunk 3]
```diff
@@ -58,6 +61,24 @@ def __init__(self, parent):
             # NB : we use a default parameter to force the evaluation of the binding
             self.actionsMap[escSeq] = lambda bindedColor=wxcolor: self.BeginTextColour(bindedColor)
 
+    def PrepocessAndWriteText(self, content):
+        """Write text into console, while capturing URLs and making 
+        them blue, underlined, and clickable.
+        """
+        l=re.split(self.regex_urls, content)
+        while l:
```
Comment (chriskiehl) [inline gooey/gui/components/widgets/richtextconsole.py]: I still feel like you're over thinking this a bit with the while loop + manual deletions of list items as you consume them. Iterators can handle all this for you (without the hefty O(n) tax for those list deletions) 

```
    textStream = iter(re.split(r, x))
    for item in textStream:
        plaintext = item
        url = next(textStream, None)
        # no additional bookkeeping required
        write(plaintext)
        if url:
            write(url)
```

Comment (eladeyal-intel) [reply gooey/gui/components/widgets/richtextconsole.py]: done. also added mailto: URLs while at it

**Comment by chriskiehl** [comment]
👍

**Pull request merged**